### ISSUE TITLE
[Kommentierung 1.5.3] - A_23541 - Servicelokalisierung durch das Clientmodul

### ISSUE THREAD

**Issue by CLAM01** [gematik/api-kim#38]
A_23541 - Servicelokalisierung durch das Clientmodul
Das Clientmodul MUSS den FQDN des Fachdienstes sowie den zu nutzenden Port per DNS 94 Service Discovery bestimmen, wenn diese nicht durch das Clientsystem im jeweiligen Benutzernamen bereitgestellt wurden.[<=]

In der Beschreibung steht, dass das CM den FQDN und den Port des Fachdiensten per DNS SD ermitteln muss wenn der Teilstring fehlt.
Soweit OK.

Jedoch müssen diese Afo nochmal geprüft werden durch gematik, da diese nachfolgende Afo besagt, dass der Vorgang abgebrochen werden muss wenn der Benutzername nicht passt im Aufbau:
Bevor ich feststellen kann das was fehlt, muss ja zuerst die Prüfung der Benutzernamen stattfinden und das CM käme nie bis zur A_23541.

A_21519-02 - Überprüfung des SMTP-Benutzernames
Das Clientmodul MUSS die übergebene SMTP-Benutzername-Zeichenkette auf Vollständigkeit überprüfen. Werden optionale Bestandteile des SMTP-Benutzernamens nicht genutzt, MUSS sichergestellt werden, dass später folgende optionale Bestandteile in ihrer vorgegebenen Position platziert werden. Als Platzhalter ist in so einem Fall "*" zu verwenden. Wenn die SMTP-Benutzername-Zeichenkette nicht vollständig ist, MUSS das Clientmodul den SMTP Fehlercode gemäß Tabelle "Tab_SMTP_Verbindung SMTP-Antwortcodes für MTA-Verbindungsaufbau" an das Clientsystem senden und den Vorgang abbrechen.

A_21518-02 - Überprüfung des POP3-Benutzernames
Das Clientmodul MUSS die übergebene POP3-Benutzername-Zeichenkette auf Vollständigkeit überprüfen. Werden optionale Bestandteile des POP3-Benutzernamens nicht genutzt, MUSS sichergestellt werden das später folgende optionale Bestandteile in ihrer vorgegebenen Position platziert werden. Als Platzhalter ist in so einem Fall "*" zu verwenden. Wenn die POP3-Benutzername-Zeichenkette nicht vollständig ist, MUSS das Clientmodul den POP3 Fehlercode gemäß Tabelle "Tab_POP3_Verbindung Antwortcodes für POP3-Server-Verbindungsaufbau" an das Clientsystem senden und den Vorgang abbrechen. [<=]

**Comment by dhufnagel** [comment]
Was ist der Sinn hinter der Ergänzung des SMTP und POP3 Nutzernamen um die Platzhalter „*“? Warum sollte man das machen, wenn das einzige System, welches diese Teile auswertet das Clientmodul selbst ist? Der Nutzername wird so nie an den Fachdienst übermittelt oder anderweitig verwendet. Daher erschließt sich mir der Sinn des Teils der jeweiligen AFO nicht.

**Comment by CLAM01** [comment]
Die "*" sind notwendig, da z.B. bei POP3 und der Nutzung einer HBA eine "User-ID" im Benutzernamen benötigt wird.
Bei SMTP hingegen gibt es die "User-ID" nicht, da nicht auf das Schlüsselmaterial des HBA zugegriffen werden muss.
Um auf das Schlüsselmaterial des HBA zugreifen zu können, muss diese ja in eine erhöhte Sicherheitsstufe gelegt werden.

Zudem befindet sich im Benutzernamen ja auch noch die KonnektorID, mit der man steuern kann, welcher Konnektor für Signatur und Verschlüsselung der KIM-Nachricht genutzt werden soll.

Würde man jetzt bei POP3 nicht das "*" bei UserId setzten, dann würde das CM die KonektorId als UserId ansehen und der Konnektor könnte nicht explizit angesteuert werden.

**Comment by dhufnagel** [comment]
Das CM muss doch auch die "*" an der gegebenen Stelle als "nicht gesetzt" interpretieren, was kein Unterschied dazu ist, ob der Teilstring einfach nicht existiert. Technisch würde ich im CM in einer Zeile den Benutzername so anpassen, dass "*" drin steht, 2 Zeilen weiter unten würde ich den Benutzername auseinandernehmen und müsste das "*" wieder interpretieren. Dann kann ich mir die Ergänzung sparen und einfach die Abfrage, ob eine User-ID angegeben ist oder nicht, anhand des Originalstrings machen und auswerten, dass ein Teil fehlt. 
Daher verstehe ich nicht, warum das Ergänzen des Benutzernamens eine AFO ist. Die AFO müsste eher fachlicher Natur sein und jedem ist es dann überlassen, wie er das implementiert.

**Comment by CLAM01** [comment]
Das CM bekommt den Benutzernamen doch vom Clientsystem.
Wenn das CM die Kombination aus hrst_domain.kim.telematik:Port nicht finden, dann muss es DNS SD machen um den FQDN zu bekommen.

Zu den optionalen Parametern.
Laut Spezifikation muss das "*" nui inmitten des Benutzernamens gesetzt werden.
Wenn ein optionaler Parameter am Ende weggelassen wird, dann muss kein "*" hin.

Bei SMTP kein Problem, da gibt es nur den optionalen Parameter KonektorId am Ende.

![image](https://github.com/gematik/api-kim/assets/115494634/019cf777-4d60-4ddf-9ccb-3003fc1ef5a8)

Bei POP3 muss das CM genau wissen wenn die UserId nicht benötigt wird.
Ok, auch wenn das "*" weggelassen wird, interpretiert das CM den POP3 Benutzernamen als OK, aber es kann dan zu Problemen kommen wenn eine HBA ins Spiel kommt.

![image](https://github.com/gematik/api-kim/assets/115494634/8313ea88-8743-47b2-8657-3c27b4b6c1c7)


Ich sehe an dieser Stelle noch etwas anders kritisch im Bezug auf die A_23541:
![image](https://github.com/gematik/api-kim/assets/115494634/000445fd-dfa5-4841-a6be-8b578235ada3)

Es werden andere Delimiter erlaubt, jedoch nicht gesagt welche.
Wenn jetzt aber der Delimiter dazu genutzt wird, den Part gemäß A_23541 zu prüfen, muss man auch wissen welcher Delimiter vorhanden sein kann. Oder?

**Comment by dhufnagel** [comment]
Ja, genau, der Benutzername kommt vom Clientsystem und nicht vom Clientmodul. Daher müsste das Clientsystem ein "\*" einfügen, nicht aber das Clientmodul, da dieses den Benutzerstring ja direkt wieder verarbeitet. Wenn die Daten vom Clientsystem fehlen, nutzt es nichts, wenn das Clientmodul vor Verarbeitung ein "\*" einfügt.

**Comment by dhufnagel** [comment]
Auf das Clientsystem hat die Spezifikation keinen Einfluss, da dies außerhalb des Regelungsbereichs der Spezifikation liegt. Die AFO beginnt mit "Das Clientmodul MUSS", daher interpretiere ich das Ergänzen als AFO für das Clientmodul.

In den anderen Punkten stimme ich voll zu.

Die Handhabung mit dem Part FQDN:Port müsste auf jeden Fall in der AFO nochmal überdacht werden. Aktuell werden alle Clientsysteme den Benutzernamen mit einem FQDN:Port Part versehen haben. Somit würde die DNS SD für bestehende, eingerichtete Clientsysteme nie ausgeführt werden, auch wenn das CM aktualisiert wird. 

Auch der Part mit dem Delimiter müsste genauer spezifiziert werden, denn hier müssten alle CM die gleichen Delimiter unterstützen, um Interoperabel zu agieren.

**Comment by gem-cp** [comment]
OK. Schauen wir uns an.

**Comment by dhufnagel** [comment]
### Zu "*" als Platzhalter
Aus meiner Sicht macht es auch Sinn, Platzhalter für die Clientsysteme zu definieren, die referenzierten AFOs interpretiere ich jedoch so, dass das Clientmodul den Benutzerstring im Clientmodul selbst mit Platzhaltern erweitern muss. Das macht aus meiner Sicht keinen Sinn. Dass das Clientmodul mit Platzhaltern umgehen können muss, macht hingegen Sinn.

**Comment by gem-cp** [comment]
OK, Wir werden die Formulierung in A_21518_XX und A_21519_XX so ändern, dass klar ist, dass das CM keine "*" Platzhalter einfügen muss. Das war nie beabsichtigt.

**Comment by stophane** [comment]
closed: wurde entsprechend angepasst, siehe ff. https://gemspec.gematik.de/docs/gemSpec/gemSpec_CM_KOMLE/gemSpec_CM_KOMLE_V1.21.0/#3.3.2.2
